utils/vlog/vlog_to_pbtype.py

```diff
@@ -54,68 +54,225 @@
 
 sys.path.insert(0, "..")
 from lib import xmlinc
+from sdf_timing import sdfparse
 
-parser = argparse.ArgumentParser(
-    description=__doc__.strip(), formatter_class=argparse.RawTextHelpFormatter
-)
-parser.add_argument(
-    'infiles',
-    metavar='input.v',
-    type=str,
-    nargs='+',
-    help="""\
-One or more Verilog input files, that will be passed to Yosys internally.
-They should be enough to generate a flattened representation of the model,
-so that paths through the model can be determined.
-"""
-)
-parser.add_argument(
-    '--top',
-    help="""\
-Top level module, will usually be automatically determined from the file name
-%.sim.v
-"""
-)
-parser.add_argument(
-    '--includes',
-    help="""\
-Command seperate list of include directories.
-""",
-    default=""
-)
-parser.add_argument('-o', help="""\
-Output filename, default 'model.xml'
-""")
 
-args = parser.parse_args()
-iname = os.path.basename(args.infiles[0])
+def make_timings(
+        pb_type_xml, sdf_file_name, sdf_cell_paths=None, sdf_variant="slow"
+):
+    """
+    Loads timings from a given SDF file and converts them to appropriate
+    statements within pb_type XML.
 
-outfile = "pb_type.xml"
-if "o" in args and args.o is not None:
-    outfile = args.o
+    If sdf_cell_name is None then the function tries to find in SDF cell with
+    name equal to the pb_type name.
 
-yosys.run.add_define("PB_TYPE")
-if args.includes:
-    for include in args.includes.split(','):
-        yosys.run.add_include(include)
+    If sdf_inst_name is None and there is only one instance of that cell in SDF
+    then that one is taken.
+    """
 
-vjson = yosys.run.vlog_to_json(args.infiles, flatten=False, aig=False)
-yj = YosysJSON(vjson)
+    def parse_timescale(timescale_str):
+        """
+        Parses a timescale expression and returns its numerical value
+        """
+        match = re.match("([0-9]+)(fs|ps|ns|us|ms|s?)$", timescale_str)
+
+        base = float(match.group(1))
+        suffix = match.group(2)
+
+        if suffix == "fs":
+            base *= 1e-15
+        elif suffix == "ps":
+            base *= 1e-12
+        elif suffix == "ns":
+            base *= 1e-9
+        elif suffix == "us":
+            base *= 1e-6
+        elif suffix == "ms":
+            base *= 1e-3
+
+        return base
+
+    def collect_timings(sdf, sdf_cell_paths):
+        """
+        Collects timings for multiple cells which paths are given in the path
+        list.
+        """
+
+        collected_timings = {}
+
+        # For each path
+        for sdf_cell_path in sdf_cell_paths:
+
+            # Only cell name given
+            if "/" not in sdf_cell_path:
```

Comment:
in general, you should get the separator from sdf header

Reply:
Done.